account_manager/src/validator/exit.rs

```diff
@@ -0,0 +1,440 @@
+async fn get_validator_index<E: EthSpec>(
+    client: &BeaconNodeHttpClient,
+    validator_pubkey: &PublicKey,
+    epoch: Epoch,
+    spec: &ChainSpec,
+) -> Result<u64, String> {
+    let validator_data = client
+        .get_beacon_states_validator_id(
+            // TODO: verify this is the state that we want to query.
+            StateId::Slot(epoch.start_slot(E::slots_per_epoch())),
+            &ValidatorId::PublicKey(validator_pubkey.into()),
+        )
+        .await
+        .map_err(|e| format!("Failed to get validator details: {:?}", e))?
+        .ok_or_else(|| {
+            format!(
+                "Validator {} is not present in the beacon state. Please ensure that your beacon node is synced",
+                validator_pubkey
+            )
+        })?
+        .data;
+
+    match validator_data.status {
+        ValidatorStatus::Active => {
+            let eligible_epoch =
+                validator_data.validator.activation_epoch + spec.shard_committee_period;
+            if epoch >= eligible_epoch {
+                Ok(validator_data.index)
+            } else {
+                Err(format!(
+                    "Validator {:?} is not eligible for exit. It will become eligible on epoch {}",
+                    validator_pubkey, eligible_epoch
+                ))
+            }
+        }
+        status => Err(format!(
+            "Validator {:?} is not eligible for voluntary exit. Validator status: {:?}",
+            validator_pubkey, status
+        )),
+    }
+}
+
+/// Get genesis data by querying the beacon node client.
+async fn get_geneisis_data(client: &BeaconNodeHttpClient) -> Result<GenesisData, String> {
+    Ok(client
+        .get_beacon_genesis()
+        .await
+        .map_err(|e| format!("Failed to get beacon genesis: {}", e))?
+        .data)
+}
+
+/// Get fork object for the current state by querying the beacon node client.
+async fn get_beacon_state_fork<E: EthSpec>(
+    client: &BeaconNodeHttpClient,
+    epoch: Epoch,
+) -> Result<Fork, String> {
+    Ok(client
+        // TODO: verify this is the state that we want to query.
+        .get_beacon_states_fork(StateId::Slot(epoch.start_slot(E::slots_per_epoch())))
+        .await
+        .map_err(|e| format!("Failed to get get fork: {:?}", e))?
+        .ok_or_else(|| "Failed to get fork, state not found".to_string())?
+        .data)
+}
+
+/// Calculates the current epoch from the genesis time and current time.
+fn get_current_epoch<E: EthSpec>(genesis_time: u64, spec: &ChainSpec) -> Result<Epoch, ArithError> {
+    let current_time = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("UNIX EPOCH should always be lower than current time")
+        .as_secs();
+
+    let elapsed = current_time.safe_sub(genesis_time)?;
+    let seconds_per_slot = spec.milliseconds_per_slot.safe_div(1000)?;
+
+    let current_slot = Slot::new(elapsed.safe_div(seconds_per_slot)?);
```

Comment:
This doesn't consider the case where `GENESIS_SLOT > 0`, which whilst unlikely here, will happen for shard chains. The safest thing to do here is to start a `SystemTimeSlotClock` and read it, like in the tests.

Reply:
Thank you. That's much better :)